### ISSUE TITLE
Implement 2d mirror error files

### ISSUE THREAD

**Issue by moellep** [radiasoft/sirepo#65]
Allow 2d mirror error files to be used in the beamline.

**Comment by mrakitin** [comment]
There are 2 users at NSLS-II interested in this functionality.

**Comment by mrakitin** [comment]
We will have an experiment next week, and need this functionality before that.

**Comment by mrakitin** [comment]
1D profile:
```python
    #HDM (Height Profile Error)
    if('HDM' in arElNames): 
        horApHDM = 0.94e-03 #Projected dimensions
        verApHDM = 1.e-03
        angHDM = 3.1415926e-03 #? grazing angle
        ifnHDM = os.path.join(_v.fdir, _v.op_HDM_ifn) if len(_v.op_HDM_ifn) > 0 else ''
        if(len(ifnHDM) > 0):
            hProfDataHDM = srwl_uti_read_data_cols(ifnHDM, '\t', 0, 1)
            opHDM = srwl_opt_setup_surf_height_1d(hProfDataHDM, 'x', _ang=angHDM, _amp_coef=_v.op_HDM_amp, _nx=1000, _ny=200, _size_x=horApHDM, _size_y=verApHDM, _xc=_v.op_HDM_x, _yc=_v.op_HDM_y)
            ofnHDM = os.path.join(_v.fdir, _v.op_HDM_ofn) if len(_v.op_HDM_ofn) > 0 else ''
            if(len(ofnHDM) > 0):
                pathDifHDM = opHDM.get_data(3, 3)
                srwl_uti_save_intens_ascii(pathDifHDM, opHDM.mesh, ofnHDM, 0, ['', 'Horizontal Position', 'Vertical Position', 'Opt. Path Dif.'], _arUnits=['', 'm', 'm', 'm'])
            el.append(opHDM); pp.append(_v.op_HDM_pp)
```
-
2D profile:
```python
        #Height Profile Error
        ifnHFM = os.path.join(_v.fdir, _v.op_HFM_ifn) if len(_v.op_HFM_ifn) > 0 else ''
        if(len(ifnHFM) > 0):
            #hProfDataHFM = srwl_uti_read_data_cols(ifnHFM, '\t', 0, 1)
            hProfDataHFM = srwl_uti_read_data_cols(ifnHFM, '\t')
            opHFM = srwl_opt_setup_surf_height_2d(hProfDataHFM, 'x', _ang=_v.op_HFM_ang, _amp_coef=_v.op_HFM_amp, _nx=1500, _ny=200)
            ofnHFM = os.path.join(_v.fdir, _v.op_HFM_ofn) if len(_v.op_HFM_ofn) > 0 else ''
            if(len(ofnHFM) > 0):
                pathDifHFM = opHFM.get_data(3, 3)
                srwl_uti_save_intens_ascii(pathDifHFM, opHFM.mesh, ofnHFM, 0, ['', 'Horizontal Position', 'Vertical Position', 'Opt. Path Dif.'], _arUnits=['', 'm', 'm', 'm'])
            el.append(opHFM); pp.append(_v.op_HFMT_pp)
```

**Comment by mrakitin** [comment]
Here are the results for the calculations with the 1D versus 2D profiles applied to the ellipsoidal mirror:

Orientation: | 1D: | 2D:
:-------------------------:|:-------------------------:|:-------------------------:
Horizontal | ![1d-horizontal](https://cloud.githubusercontent.com/assets/13209176/17571644/17285370-5f1f-11e6-8930-fae7aa4fb051.png) | ![2d-horizontal](https://cloud.githubusercontent.com/assets/13209176/17571643/171f4ece-5f1f-11e6-9b76-30bffe95dbdd.png)
Vertical | ![1d-vertical](https://cloud.githubusercontent.com/assets/13209176/17571796/96297df2-5f1f-11e6-918e-07683e9d8380.png) | ![2d-vertical](https://cloud.githubusercontent.com/assets/13209176/17571800/9e084e36-5f1f-11e6-8e99-142a3c305b36.png)

**Comment by mrakitin** [comment]
Implemented in 0fba247 for:
- crystal:

No profile | 2D profile
:-:|:-:
![crystal no profile](https://cloud.githubusercontent.com/assets/13209176/17572918/1e82acaa-5f25-11e6-9a33-5f429dcaafd5.png) | ![crystal 2d profile](https://cloud.githubusercontent.com/assets/13209176/17572926/24168272-5f25-11e6-9d2f-8237ba866ba3.png)

- ellipsoid mirror:

No profile | 2D profile
:-:|:-:
![ellipsoid mirror no profile](https://cloud.githubusercontent.com/assets/13209176/17572963/51d57150-5f25-11e6-92a2-4e13763daf38.png) | ![ellipsoid mirror 2d profile](https://cloud.githubusercontent.com/assets/13209176/17572969/562046f4-5f25-11e6-9de7-ae649565bd8e.png)

- mirror:

No profile | 1D profile | 2D profile
:-:|:-:|:-:
![mirror no profile](https://cloud.githubusercontent.com/assets/13209176/17572989/717faca0-5f25-11e6-9c7c-f2bc438de0d5.png) | ![mirror 1d profile](https://cloud.githubusercontent.com/assets/13209176/17573009/84a22c9a-5f25-11e6-896d-f2f5ac3bb259.png) | ![mirror 2d profile](https://cloud.githubusercontent.com/assets/13209176/17573014/8a18c7d8-5f25-11e6-884d-3f4e7af1114f.png)

- spherical mirror:

No profile | 2D profile
:-:|:-:
![spherical mirror no error](https://cloud.githubusercontent.com/assets/13209176/17573035/b667f138-5f25-11e6-90b9-5ae8f6e4943c.png) | ![spherical mirror 2d error](https://cloud.githubusercontent.com/assets/13209176/17573034/b666576a-5f25-11e6-8d33-98e5875e1ec1.png)

**Comment by mrakitin** [comment]
Merged to master. Closing.
